### PR TITLE
[Clang] Require base element type of `__has_unique_object_representations` to be complete

### LINKED ISSUE #95311: [clang] __has_unique_object_representations gives inconsistent answer based on instantiation order
The following code reports inconsistent results for `__has_unique_object_representations`:

```c++
template <int>
class Foo {
  int x;
};

static_assert(__has_unique_object_representations(Foo<0>[]));
static_assert(__has_unique_object_representations(Foo<0>[][3]));
static_assert(__has_unique_object_representations(Foo<0>));
```

The key here is that if you assert `__has_unique_object_representations(Foo<0>)` **first**, then everything works. So somehow asking for `__has_unique_object_representations(Foo<0>)` must cause something to be instantiated or stored in the AST and that changes the result of a subsequent `__has_unique_object_representations(Foo<0>[])` query (notice the array).

This bug was extracted from this comment: https://github.com/llvm/llvm-project/pull/69241#issuecomment-2163612863

Godbolt: https://gcc.godbolt.org/z/Pb9ze7YxM

### PR DESCRIPTION
**Pull request by MitalAshok**
Fixes #95311

Previous behaviour was that `false` was silently returned, templated classes were not instantiated and incomplete classes did not issue an error.

### CODE REVIEW

File: clang/lib/AST/ASTContext.cpp [hunk 1]
```diff
@@ -2791,6 +2791,12 @@ bool ASTContext::hasUniqueObjectRepresentations(
     return hasUniqueObjectRepresentations(getBaseElementType(Ty),
                                           CheckIfTriviallyCopyable);
 
+  if (Ty->isVoidType())
```
Comment (shafik) [inline clang/lib/AST/ASTContext.cpp:2794]: It is not clear why we need this change as well.

Comment (MitalAshok) [reply clang/lib/AST/ASTContext.cpp:2794]: For the new `!Ty->isIncompleteType()` assertion below. Before `void` would go to the `!isTriviallyCopyable` path, this is just a more explicit way of handling `__has_unique_object_representations(void)`. The assertion could also be `assert(Ty->isVoidType() || !Ty->isIncompleteType())` to follow the spec better?

Comment (cor3ntin) [reply clang/lib/AST/ASTContext.cpp:2794]: I'd prefer that, I think

**Review by cor3ntin — APPROVED** [review]
LGTM

**Pull request merged**